src/Server/HTTPHandler.cpp

```diff
@@ -1043,12 +1043,21 @@ void HTTPHandler::formatExceptionForClient(int exception_code, HTTPServerRequest
 
     /// FIXME: make sure that no one else is reading from the same stream at the moment.
 
-    /// If HTTP method is POST and Keep-Alive is turned on, we should read the whole request body
+    /// If HTTP method is POST and Keep-Alive is turned on, we should try to read the whole request body
     /// to avoid reading part of the current request body in the next request.
     if (request.getMethod() == Poco::Net::HTTPRequest::HTTP_POST && response.getKeepAlive()
-        && exception_code != ErrorCodes::HTTP_LENGTH_REQUIRED && !request.getStream().eof())
+        && exception_code != ErrorCodes::HTTP_LENGTH_REQUIRED)
     {
-        request.getStream().ignoreAll();
+        try
+        {
+            if (!request.getStream().eof())
```

Comment:
I explained this a bit in the issue (https://github.com/ClickHouse/ClickHouse/issues/65117). In case of HTTP keep-alive, as part of the exception handling, we try to drain the remaining request data from the socket "to avoid reading part of the current request body in the next request" (quoting the existing comment). 

[ReadBuffer::ignoreAll](https://github.com/ClickHouse/ClickHouse/blob/113bb0000510b30c0845593911baa6d72cd5fb20/src/IO/ReadBuffer.h#L145) (and [ReadBuffer::eof](https://github.com/ClickHouse/ClickHouse/blob/113bb0000510b30c0845593911baa6d72cd5fb20/src/IO/ReadBuffer.h#L104) as a side effect) may attempt to receive data from the socket, which may result in a timeout exception (see https://github.com/ClickHouse/ClickHouse/issues/65117). However, depending on the `ReadBuffer` implementation, these calls may also fail for other reasons. An example is the [adapted test](https://github.com/ClickHouse/ClickHouse/pull/65118/files#diff-13aef4ed7a19b67b726fd5522064bfde5726192b76a07e309cc170be87129517) where [HTTPChunkedReadBuffer](https://github.com/ClickHouse/ClickHouse/blob/2909e6451b6e0118679b2b96b1c3e26e1e226870/src/IO/HTTPChunkedReadBuffer.cpp#L36) throws if the advertised chunk size exceeds the allowed limit.